### DISCUSSION TITLE
Constructor Destructor syntax improvement

### DISCUSSION THREAD

**Discussion by lauler1** [hsutter/cppfront#1039] [Suggestions]
Hi

I'd like to propose a modification for enhancing the syntax in Cppfront, aimed purely at syntactic sugar. In line with the objective to simplify the language 10x, it seems to me that the current syntax for constructors and destructors may not fully align with this goal, particularly for those new to programming or coming from different programming backgrounds. I mean no criticism by this; it's merely a point for consideration. I'm not sure if this topic has been broached before.

The use of `out this`, `inout this`, and `move this` within the context of overloading the `operator=` appears to me as potentially confusing and not entirely transparent regarding its purpose for many programmers. To enhance clarity, an alternative could be to adopt terms like `construct this`, `assign this`, and `destruct this` respectively.

Thank you for your work.

**Comment by zaucy** [comment]
Adding additional keywords for constructor/destructor complicates the language more. I would have to memorize what `construct` translates to. Does it work for assignment as well?

Using `out this` as a constructor/assignment is natural and consistent since I can write a regular member function with the same behaviour.

**Reply by lauler1** [reply]
I just wanted to give an initial idea, the exact syntax could be better elaborated.
Constructors and destructors are concepts widely recognized by programmers across numerous languages, in contrast to the more complex notions of operator overloading or move semantics, which can be challenging for newcomers to grasp. My aim is to address a potential reluctance that might arise with Cppfront.
Curious query: Who is more likely to lead the adoption of Cppfront, novice programmers or experienced C++ developers?
Introducing keywords for constructors and destructors (that reinforce the intention) as optional alias for `out` and `move` within class definitions could offer both syntaxes, making the language more accessible.

**Reply by zaucy** [reply]
> complex notions of operator overloading or move semantics, which can be challenging for newcomers to grasp

`operator=` can be a gentle introduction to this concept. Function overloading is a feature of cpp that isn't going away. In my opinion, changing the rules for constructor/destructor to hide function overloading would hinder learning rather than help it.

Even in C++ today if you declare a copy constructor there is an implicit `operator=` that gets generated. This behaviour has to be taught or worse the developer is blissfully unaware. In `cpp2` utilising the `operator=` directly is more clear and consistent. Especially since constructing (with parameters) always involves the `=` symbol.

**Comment by lauler1** [comment]
I understand
